### PR TITLE
VFB-170-Fix role not found

### PR DESCRIPTION
**Pull request by minalui**
## What's changed
Handle role not found errors

## Screenshots / Videos
| Before     | After      |
|------------|------------|
| paste_here | paste_here |

## Checklist
- [ ] The ticket is up-to-date - Please document any deviations from the original approach if there is any.
- [ ] I have documented the testing steps for QA
- [x] I have self-reviewed this PR
- [x] Make sure you've verified it works via `npm run dev`
- [x] Make sure you've verified it works via `npm run build` and `npm run start`
- [x] Make sure you've fixed all linting problems with `npm run lint_fix`
- [x] Make sure you've tested via `npm run test:e2e`

If you have made any changes to the database... N/A

  - [ ] The migration files are up-to-date with my final set up (`npx supabase db diff -f <name_of_migration>` should create nothing at this point)
  - [ ] I have updated the typescript definitions for the database with `db:local:generate_types`
  - [ ] I have modified the seed in `supabase/seed/seed.mts` if appropriate
  - [ ] If I have modified the seed, I have also generated the seed with `npm run db:generate_seed` 
  - [ ] With my final set up, I can run `npm run dev:reset_supabase` without any errors.
  - Does this require resetting the deployed database? - YES / NO (remove as appropriate)
  - [ ] I have checked that migration can happen successfully without resetting the database, doing the following.
    - Make sure you have rebased your branch onto dev or merged dev into your branch.
    - Checkout the dev branch
    - Run `npm run post_checkout` to reset the database, including the seed data.
    - Checkout your branch
    - Run `npx supabase migrations up --include-all --local` to run all outstanding migration files
    - Check that the resulting database is what you expect.

### CODE REVIEW

File: src/app/admin/page.tsx [hunk 1]
```diff
@@ -31,9 +30,20 @@ const getUsers = async (): Promise<UserRow[]> => {
...
+
+        const { role, error } = await fetchUserRole(user.id);
+
+        if (error) {
+            userRole = "UNKNOWN";
+            const logId = await logErrorReturnLogId(
+                `failed to fetch user role. User ID: ${user.id}`,
+                error
+            );
+            throw new DatabaseError("fetch", "user role for users table", logId);
```
Comment (minalui) [inline src/app/admin/page.tsx:43]: If thrown, then userRole = "UNKNOWN" will not be passed through

File: src/server/authenticateAdminUser.ts [hunk 2]
```diff
@@ -36,9 +37,16 @@ export async function authenticateAsAdmin(): Promise<Authenticated> {
...
-        return logErrorReturnLogId("Failed to fetch user role", error);
-    });
+    const { role: userRole, error } = await fetchUserRole(user.id);
+
+    if (error) {
+        const logId = await logErrorReturnLogId(
+            `Failed to fetch user role to authenticate as admin ${user.id}`,
+            error
+        );
+        throw new DatabaseError("fetch", "user role for admin authorisation", logId);
```
Comment (minalui) [inline src/server/authenticateAdminUser.ts:47]: Should we just return an error?

File: src/app/admin/manageUser/EditUserForm.tsx [hunk 3]
```diff
@@ -18,8 +18,16 @@ interface Props {
...
+    return userRole !== "UNKNOWN";
+}
+
 const EditUserForm: React.FC<Props> = (props) => {
-    const [role, setRole] = useState(props.userToEdit.userRole);
+    const initialRole = isValidUserRole(props.userToEdit.userRole)
+        ? props.userToEdit.userRole
+        : "caller";
+
+    const [role, setRole] = useState(initialRole);
```
Comment (rickaw) [inline src/app/admin/manageUser/EditUserForm.tsx:30]: Let's specify type here to make it difficult to break logic around here
```suggestion
    const initialRole: UserRole = isValidUserRole(props.userToEdit.userRole)
        ? props.userToEdit.userRole
        : "caller";

    const [role, setRole] = useState<UserRole>(initialRole);
```

**Review by rickaw — APPROVED** [review]
Looks great, a few tiny comments!

File: src/app/admin/page.tsx [hunk 4]
```diff
@@ -5,19 +5,19 @@ import AdminPage from "@/app/admin/AdminPage";
...
+import { Schema, UserRole } from "@/databaseUtils";
 import { logErrorReturnLogId } from "@/logger/logger";
-import { Database } from "@/databaseTypesFile";
 import { adminGetUsers } from "@/server/adminGetUsers";
 import { fetchUserRole } from "@/common/fetchUserRole";
 
 // disables caching
 export const revalidate = 0;
 
+export type DisplayedUserRole = UserRole | "UNKNOWN";
```
Comment (rickaw) [inline src/app/admin/page.tsx:16]:
```suggestion
export type DisplayedUserRole = UserRole | "UNKNOWN";

```

**Review by rickaw — APPROVED** [review]
Nice! Feel free to merge

**Pull request merged**